### DISCUSSION TITLE
How to add tracing filter processor ?

### DISCUSSION THREAD

**Discussion by hadashiA** [open-telemetry/opentelemetry-dotnet#3201] [Q&A]
I would like to filter Tracing based on Activity instance.
I have looked here and attempted to create a filter processor for Tracing.

https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/README.md#filtering-processor

However, I have two problems.

1. When decorating Processor as follows.  `SetParentProvider` of `SimpleActivityExportProcessor` will not be called. So Exporter will not reference ResourceBuilder.
    - `new MyFilteringProcessor( new SimpleActivityExportProcessor(new MyExporter("ExporterX"))`
    - refs: https://github.com/open-telemetry/opentelemetry-dotnet/blob/2a97920ff0a603837a1121204824955bba739c57/src/OpenTelemetry/Trace/TracerProviderSdk.cs#L264
3. Since `AddOtlpExporter` like extension methods has a hard-coded ExportProcessor, it is not possible to insert a Processor without re-implementing the entire extension method itself.
    - https://github.com/open-telemetry/opentelemetry-dotnet/blob/18d73c59f021b1c7a9bd433ea703e6f2c22a0b90/src/OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs/OtlpLogExporterHelperExtensions.cs

I am particularly troubled by 1. Can anyone tell me a good way to add a filter? Thanks.

**Comment by alanwest** [comment]
Hmm, this is the first I've taken a look at the [example you reference](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/README.md#filtering-processor).

Personally, I don't think a processor is a great solution for doing this. I'd be more inclined to write my own exporter that does the filtering. For example, you could write an exporter that wrapped the OTLP exporter. Your exporter would do the filtering and forward to the underlying exporter.

Regarding `SetParentProvider` not getting invoked - yes I agree this is a problem with the example as is. Though I think replacing this example with an example of a custom exporter that does filtering may be a better solution than trying to provide a mechanism to set the provider.

**Reply by hadashiA** [reply]
Thanks for the reply.

Hmmm. I understand.
 I feel that the "processors" -> "exporters" pipeline is intended to increase portability without changing each exporter, etc. Therefore, I thought it would be more flexible to do filter processing independent of the exporter implementation at the previous stage.

But anyway, I understand that this is not possible. Thanks!

**Comment by alanwest** [comment]
@hadashiA This was recently fixed in #3368. The example needed to be updated as well. https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/docs/trace/extending-the-sdk/MyFilteringProcessor.cs
